pkg/core/blockchain.go

```diff
@@ -430,28 +430,26 @@ func (bc *Blockchain) AddBlock(block *block.Block) error {
 		if err != nil {
 			return fmt.Errorf("block %s is invalid: %w", block.Hash().StringLE(), err)
 		}
-		if bc.config.VerifyTransactions {
-			var mp = mempool.New(len(block.Transactions))
-			for _, tx := range block.Transactions {
-				var err error
-				// Transactions are verified before adding them
-				// into the pool, so there is no point in doing
-				// it again even if we're verifying in-block transactions.
-				if bc.memPool.ContainsKey(tx.Hash()) {
-					err = mp.Add(tx, bc)
-					if err == nil {
-						continue
-					}
-				} else {
-					err = bc.verifyAndPoolTx(tx, mp)
-				}
-				if err != nil {
-					return fmt.Errorf("transaction %s failed to verify: %w", tx.Hash().StringLE(), err)
+		mp = mempool.New(len(block.Transactions))
+		for _, tx := range block.Transactions {
+			var err error
+			// Transactions are verified before adding them
+			// into the pool, so there is no point in doing
+			// it again even if we're verifying in-block transactions.
+			if bc.memPool.ContainsKey(tx.Hash()) {
+				err = mp.Add(tx, bc)
+				if err == nil {
+					continue
 				}
+			} else {
+				err = bc.verifyAndPoolTx(tx, mp)
+			}
+			if err != nil && bc.config.VerifyTransactions {
+				return fmt.Errorf("transaction %s failed to verify: %w", tx.Hash().StringLE(), err)
```

Comment:
Yeah, that's the most problematic commit. But without this change we're still creating a sorted list of transactions for subsequent searches. What's the difference between VT on and off for the current master?

Reply:
The only problem performance-wise here is that mempool incurs a bit more overhead than a sorted list, but that could be solved for cases like this one (the other one is in the consensus). Other than that we're either building a map or building a list, one way of another we need some efficient structure for `isTxStillRelevant`.

Reply:
The average TPS values:
| Configuration | VT = true | VT = false |
| --- | --- | --- |
| Single, 30 wrk, master | 10095.744 | 9896.571 |
| Single, 30 wrk, tps-part-3 | 10073.993 | 10479.104 |

Avg TPS with disabled validation from master branch is lower then the one with validation enabled, which is strange a bit.

Reply:
I guess it'd be correct to say that the difference now is lower than the measurement error which means that the change can be applied without any significant losses (and because it slightly improves the default VT-enabled case).